### PR TITLE
Feat/129 pennant chart

### LINKED ISSUE #129: Add pennant chart
- Render pennant chart

### PR DESCRIPTION
**Pull request by johnwalley**
Closes #129.

### Depth chart uses the update subscription

This means it receives a delta rather than the full market depth with each message. However, there are some oddities:

- it appears to return data for more than just the market you asked for
- it appears to do the above on a round robin basis which leads to periods of time where the chart does not update
- the web socket connection is closed after 30s (we reconnect straight away but it's not ideal)
- most importantly we do not seem to receive every message (non monotonic sequence numbers)

### Dependency on `jest-canvas-mock'`

The charts make heavy use of the HTML canvas element. This dependency mocks out the canvas element for Jest tests.

### Noisy log messages

When building or serving the trading app pennant is generating some unwanted log messages due to running outside a browser environment.

Screenshot 2022-04-06 at 18 04 00

I will remove these in pennant in the next release.

### CODE REVIEW

**Review by mattrussell36 — CHANGES_REQUESTED** [review]
A few minor things to update here to keep things consistent with otherfiles/components in the mono repo.

Additionally, I'm wondering if we should/could refactor this later on to use our data provider pattern which we use for other high frequency data components (trades tables, markets tables etc). Definitely an iteration though!

**Comment by johnwalley** [comment]
> Additionally, I'm wondering if we should/could refactor this later on to use our data provider pattern which we use for other high frequency data components (trades tables, markets tables etc). Definitely an iteration though!

I discussed this with Bart this morning and I think we both understand what would need to be done to use it for the depth chart and order list components.

File: apps/trading/components/chart-container/chart-container.tsx [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    <div style={{ display: 'flex', flexDirection: 'column', height: '100%' }}>
+      <ButtonRadio
+        name="chart-type"
+        options={[
+          { text: 'Trading view', value: 'trading' },
+          { text: 'Depth', value: 'depth' },
+        ]}
+        currentOption={chartType}
+        onSelect={(value) => {
+          setChartType(value as 'depth' | 'trading');
```
[the remark this answers is not in the document]
Comment (johnwalley) [reply apps/trading/components/chart-container/chart-container.tsx]: I've made the ButtonRadio component generic in its props.

File: libs/chart/src/lib/data-source.ts [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    this.candlesSub && this.candlesSub.unsubscribe();
+  }
+
+  /**
+   * Used by the charting library to create a subscription to streaming annotation data.
+   */
+  subscribeAnnotations(
+    onSubscriptionAnnotation: (annotations: Annotation[]) => void
+  ) {
+    console.log('subscribeAnnotations');
```
[the remark this answers is not in the document]
Comment (johnwalley) [reply libs/chart/src/lib/data-source.ts]: Turns out it was optional all along. Removed.

File: libs/depth-chart/src/lib/hooks/use-depth-update.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+} from '../__generated__/marketDepthUpdateSubscribe';
+
+export interface QueryResult<TData> {
+  data: TData | undefined;
+  loading: boolean;
+  error?: ApolloError;
+}
+
+export function useDepthUpdate({ marketId }: marketDepthVariables, wait = 0) {
+  const queryResultRef = React.useRef<QueryResult<marketDepth>>({
```
[the remark this answers is not in the document]
Comment (johnwalley) [reply libs/depth-chart/src/lib/hooks/use-depth-update.ts]: Updated all React imports to be consistent.

**Review by mattrussell36 — APPROVED** [review]

File: apps/trading/components/chart-container/button-radio.tsx [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import { Button } from '@vegaprotocol/ui-toolkit';
+
+interface ButtonRadioProps<T> {
+  name: string;
+  options: Array<{ value: T; text: string }>;
+  currentOption: T | null;
+  onSelect: (option: T) => void;
+}
+
+export const ButtonRadio = <T extends string>({
```
Comment (dexturr) [inline apps/trading/components/chart-container/button-radio.tsx:10]: follow up (non-blocking): can we move this to or use something from ui toolkit?

**Review by dexturr — APPROVED** [review]

File: libs/depth-chart/src/lib/hooks/use-depth-update.ts [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  marketDepthUpdateSubscribeVariables,
+} from '../__generated__/marketDepthUpdateSubscribe';
+
+export interface QueryResult<TData> {
+  data: TData | undefined;
+  loading: boolean;
+  error?: ApolloError;
+}
+
+export function useDepthUpdate({ marketId }: marketDepthVariables, wait = 0) {
```
Comment (bglownia) [inline libs/depth-chart/src/lib/hooks/use-depth-update.ts:25]: We may want to switch to makeDataProvider
https://github.com/vegaprotocol/frontend-monorepo/blob/feature/151-orderbook/libs/orderbook/src/lib/marketdepth-data-provider.ts

Comment (johnwalley) [reply libs/depth-chart/src/lib/hooks/use-depth-update.ts:25]: I've created #226 for this task.

File: libs/chart/src/lib/data-source.ts [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    Interval.I5M,
+    Interval.I1M,
+  ],
+  priceMonitoringBounds: [],
+};
+
+/**
+ * A data access object that provides access to the Vega GraphQL API.
+ */
+export class VegaDataSource implements DataSource {
```
[the remark this answers is not in the document]
Comment (johnwalley) [reply libs/chart/src/lib/data-source.ts:87]: Created #227

**Pull request merged**